test/realtime/message.test.js

```diff
@@ -1183,5 +1183,145 @@ define(['ably', 'shared_helper', 'async', 'chai'], function (Ably, helper, async
+            }
+            rtFilteredChannel.subscribe(function (msg) {
+              try {
+                // Push received filtered messages into an array
+                filteredMessages.push(msg);
+              } catch (err) {
+                closeAndFinish(done, realtime, err);
+                return;
+              }
+            });
+          });
+
+          rtUnfilteredChannel.attach(function (err) {
+            if (err) {
+              closeAndFinish(done, realtime, err);
+              return;
+            }
+
+            rtUnfilteredChannel.subscribe(function (msg) {
+              try {
+                // Push received unfiltered messages into an array
+                unFilteredMessages.push(msg);
+              } catch (err) {
+                closeAndFinish(done, realtime, err);
+                return;
+              }
+            });
+
+            // Subscription to check all messages were received as expected
+            rtUnfilteredChannel.subscribe('end', function (msg) {
+              try {
+                expect(msg.data).to.equal(testData[testData.length - 1].data, 'Unexpected msg data received');
+
+                // Check that we receive expected messages on filtered channel
+                expect(filteredMessages.length).to.equal(2, 'Expect only two filtered message to be received');
+                expect(filteredMessages[0].data).to.equal(testData[0].data, 'Unexpected data received');
+                expect(filteredMessages[1].data).to.equal(testData[2].data, 'Unexpected data received');
+                expect(filteredMessages[0].extras.headers.name).to.equal(
+                  testData[0].extras.headers.name,
+                  'Unexpected header value received'
+                );
+                expect(filteredMessages[1].extras.headers.name).to.equal(
+                  testData[2].extras.headers.name,
+                  'Unexpected header value received'
+                );
+                // Check that message with header that doesn't meet filtering condition is not received.
+                for (msg of filteredMessages) {
+                  expect(msg.extras.headers.number).to.equal(26095, 'Unexpected header filtering value received');
+                }
+
+                // Check that we receive expected messages on unfiltered channel, including the `end` event message
+                expect(unFilteredMessages.length).to.equal(6, 'Expect only 6 unfiltered message to be received');
+                for (var i = 0; i < unFilteredMessages.length; i++) {
+                  expect(unFilteredMessages[i].data).to.equal(testData[i].data, 'Unexpected data received');
+                }
+              } catch (err) {
+                closeAndFinish(done, realtime, err);
+                return;
+              }
+              closeAndFinish(done, realtime);
+            });
+            var restChannel = rest.channels.get('chan');
+            restChannel.publish(testData);
```

Comment:
Hmm, almost there. While this fixes the issue for the unfiltered channel attachment, we still have a race condition where the test might fail if the filtered channel has not yet attached before we start publishing messages. We need to ensure both attachments have succeeded before we publish :)